File: backend/routes/stats.py
import asyncio

from fastapi import APIRouter, Depends, Request
from psycopg.errors import UndefinedTable

from ..errors import api_errors
from ..security import require_auth

router = APIRouter(prefix="/api/stats", tags=["Statistics"], dependencies=[Depends(require_auth)])


@router.get("/", include_in_schema=False)
@router.get("")
async def get_stats(request: Request):
    db = request.app.state.db
    with api_errors("Failed to load stats"):
        # Fetch independent dashboard totals concurrently.
        totals, tags, cities, cases = await asyncio.gather(
            db.query("SELECT COUNT(*)::int AS count FROM persons"),
            db.query("""SELECT ct.crime_name, COUNT(DISTINCT cp.person_id)::int AS count
                        FROM cases c JOIN case_people cp ON cp.case_id=c.case_id
                        JOIN crime_types ct ON ct.crime_id = c.crime_id
                        GROUP BY ct.crime_name ORDER BY count DESC"""),
            db.query(
                "SELECT city, COUNT(*)::int AS count FROM persons GROUP BY city ORDER BY count DESC"
            ),
            db.query("SELECT COUNT(*)::int AS count FROM cases"),
        )
        traced = 0
        try:
            rows = await db.query("SELECT COUNT(*)::int AS count FROM associations")
            traced = rows[0]["count"]
        except UndefinedTable:
            # Older databases may not have the optional associations table.
            pass
        return {
            "totalCriminals": totals[0]["count"],
            "totalCases": cases[0]["count"],
            "tracedConnections": traced,
            "tagCounts": [[row["crime_name"], row["count"]] for row in tags],
            "cityCounts": [[row["city"], row["count"]] for row in cities],
        }

File: express-backend/routes/stats.js
import { Router } from "express";
import { pool } from "../db.js";

const router = Router();

// GET /api/stats — dashboard summary numbers, computed straight from the DB
router.get("/", async (req, res) => {
  try {
    const [{ rows: totalRows }, { rows: tagRows }, { rows: cityRows }, { rows: caseRows }] = await Promise.all([
      pool.query("SELECT COUNT(*)::int AS count FROM persons"),
      pool.query(`
        SELECT ct.crime_name, COUNT(DISTINCT c.person_id)::int AS count
        FROM cases c JOIN crime_types ct ON ct.crime_id = c.crime_id
        GROUP BY ct.crime_name ORDER BY count DESC
      `),
      pool.query("SELECT city, COUNT(*)::int AS count FROM persons GROUP BY city ORDER BY count DESC"),
      pool.query("SELECT COUNT(*)::int AS count FROM cases"),
    ]);

    let tracedConnections = 0;
    try {
      const { rows } = await pool.query("SELECT COUNT(*)::int AS count FROM associations");
      tracedConnections = rows[0].count;
    } catch (err) {
      if (err.code !== "42P01") throw err; // ignore "table doesn't exist yet"
    }

    res.json({
      totalCriminals: totalRows[0].count,
      totalCases: caseRows[0].count,
      tracedConnections,
      tagCounts: tagRows.map((r) => [r.crime_name, r.count]),
      cityCounts: cityRows.map((r) => [r.city, r.count]),
    });
  } catch (err) {
    console.error("GET /api/stats failed", err);
    res.status(500).json({ error: "Failed to load stats" });
  }
});

export default router;